device/cuda/src/ambiguity_resolution/greedy_ambiguity_resolution_algorithm.cu

```diff
@@ -538,33 +594,50 @@ greedy_ambiguity_resolution_algorithm::operator()(
-                .n_updated_tracks = n_updated_tracks_device.get(),
-                .is_updated_view = is_updated_buffer,
-                .block_offsets_view = block_offsets_buffer,
-                .prefix_sums_view = prefix_sums_buffer});
-
-        kernels::scan_block_offsets<<<1, nBlocks_scan,
-                                      nBlocks_scan * sizeof(int), stream>>>(
+        kernels::
+            block_inclusive_scan<<<nBlocks_scan, nThreads_scan,
+                                   nThreads_scan * sizeof(int), stream_scan>>>(
+                device::block_inclusive_scan_payload{
+                    .sorted_ids_view = sorted_ids_buffer,
+                    .terminate = terminate_device.get(),
+                    .n_accepted = n_accepted_device.get(),
+                    .n_updated_tracks = n_updated_tracks_device.get(),
+                    .is_updated_view = is_updated_buffer,
+                    .block_offsets_view = block_offsets_buffer,
+                    .prefix_sums_view = prefix_sums_buffer});
+
+        kernels::scan_block_offsets<<<
+            1, nBlocks_scan, nBlocks_scan * sizeof(int), stream_scan>>>(
             device::scan_block_offsets_payload{
                 .terminate = terminate_device.get(),
                 .n_accepted = n_accepted_device.get(),
                 .n_updated_tracks = n_updated_tracks_device.get(),
                 .block_offsets_view = block_offsets_buffer,
                 .scanned_block_offsets_view = scanned_block_offsets_buffer});
 
-        kernels::add_block_offset<<<nBlocks_scan, nThreads_scan, 0, stream>>>(
-            device::add_block_offset_payload{
-                .terminate = terminate_device.get(),
-                .n_accepted = n_accepted_device.get(),
-                .n_updated_tracks = n_updated_tracks_device.get(),
-                .block_offsets_view = scanned_block_offsets_buffer,
-                .prefix_sums_view = prefix_sums_buffer});
+        kernels::
+            add_block_offset<<<nBlocks_scan, nThreads_scan, 0, stream_scan>>>(
+                device::add_block_offset_payload{
+                    .terminate = terminate_device.get(),
+                    .n_accepted = n_accepted_device.get(),
+                    .n_updated_tracks = n_updated_tracks_device.get(),
+                    .block_offsets_view = scanned_block_offsets_buffer,
+                    .prefix_sums_view = prefix_sums_buffer});
+
+        // Record the events
+        cudaEvent_t event_main, event_fill, event_scan;
+        cudaEventCreate(&event_main);
+        cudaEventCreate(&event_fill);
+        cudaEventCreate(&event_scan);
+
+        cudaEventRecord(event_main, stream);
+        cudaEventRecord(event_fill, stream_fill);
+        cudaEventRecord(event_scan, stream_scan);
+
+        // Synchronize the events with main stream
+        cudaStreamWaitEvent(stream, event_main, 0);
+        cudaStreamWaitEvent(stream, event_fill, 0);
+        cudaStreamWaitEvent(stream, event_scan, 0);
```

Comment:
These are completely equivalent to just:

```c++
TRACCC_CUDA_CHECK(cudaStreamSynchronize(stream));
TRACCC_CUDA_CHECK(cudaStreamSynchronize(stream_fill));
TRACCC_CUDA_CHECK(cudaStreamSynchronize(stream_scan));
```

Creating events makes sense if you need to pass such events between independent code blocks. Here you just want to synchronize on the stream(s).

Reply:
`TRACCC_CUDA_ERROR_CHECK` instead of `TRACCC_CUDA_CHECK`

BTW, it is not obviously equivalent as I see a crash like the following:
```
terminate called after throwing an instance of 'vecmem::cuda::runtime_error'
  what():  /mnt/nvme0n1/byeo/projects/traccc/traccc_build/_deps/vecmem-src/cuda/src/memory/managed_memory_resource.cpp:51 Failed to execute: cudaFree(p) (operation not permitted when stream is capturing)
Aborted (core dumped)
```